File: backend/utils/logger.js
import winston from 'winston';
import 'winston-daily-rotate-file';

// Define the custom log format
const logFormat = winston.format.combine(
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
    winston.format.errors({ stack: true }),
    winston.format.printf(({ timestamp, level, message, stack }) => {
        return `[${timestamp}] ${level.toUpperCase()}: ${stack || message}`;
    })
);

// 1. General Application Transport (Info level)
const infoTransport = new winston.transports.DailyRotateFile({
    filename: 'logs/application-%DATE%.log',
    datePattern: 'YYYY-MM-DD',
    level: 'info',        // Captures info and warnings (ignores debug)
    zippedArchive: true,
    maxSize: '20m',
    maxFiles: '14d'
});

// 2. Dedicated Error Transport (Error level only)
const errorTransport = new winston.transports.DailyRotateFile({
    filename: 'logs/error-%DATE%.log',
    datePattern: 'YYYY-MM-DD',
    level: 'error',      // Exclusively captures errors
    zippedArchive: true,
    maxSize: '20m',
    maxFiles: '14d'
});

// Create the logger instance
const logger = winston.createLogger({
    level: 'info', // Sets the minimum logging level to info (disabling debug entirely)
    format: logFormat,
    transports: [
        infoTransport,
        errorTransport,
        new winston.transports.Console() // Keeps console output active for local terminal viewing
    ]
});

export default logger;

File: kyc-python-worker/worker_app/utils/logger.py
import logging
import os
from logging.handlers import TimedRotatingFileHandler

# Define the log directory (one level up from worker_app, matching backend structure)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
LOG_DIR = os.path.join(BASE_DIR, 'logs')
os.makedirs(LOG_DIR, exist_ok=True)

# Define the custom log format mirroring Winston
# Format: [YYYY-MM-DD HH:mm:ss] LEVEL: message
class CustomFormatter(logging.Formatter):
    def format(self, record):
        # Ensure level name is uppercase
        record.levelname = record.levelname.upper()
        # Format the traceback if present, similar to Winston's stack printing
        if record.exc_info:
            record.exc_text = self.formatException(record.exc_info)
            record.exc_info = None
        
        return super().format(record)

log_format = '[%(asctime)s] %(levelname)s: %(message)s'
formatter = CustomFormatter(log_format, datefmt='%Y-%m-%d %H:%M:%S')

# 1. General Application Transport (Info level)
# Mirrors Winston's application-%DATE%.log behavior
info_handler = TimedRotatingFileHandler(
    filename=os.path.join(LOG_DIR, 'application.log'),
    when='midnight',
    interval=1,
    backupCount=14,
    encoding='utf-8'
)
info_handler.setLevel(logging.INFO)
info_handler.setFormatter(formatter)
# Suffix to match daily rotation pattern
info_handler.suffix = "%Y-%m-%d.log"

# 2. Dedicated Error Transport (Error level only)
# Mirrors Winston's error-%DATE%.log behavior
error_handler = TimedRotatingFileHandler(
    filename=os.path.join(LOG_DIR, 'error.log'),
    when='midnight',
    interval=1,
    backupCount=14,
    encoding='utf-8'
)
error_handler.setLevel(logging.ERROR)
error_handler.setFormatter(formatter)
# Suffix to match daily rotation pattern
error_handler.suffix = "%Y-%m-%d.log"

# 3. Console Handler
# Mirrors Winston's new winston.transports.Console()
console_handler = logging.StreamHandler()
console_handler.setLevel(logging.INFO)
console_handler.setFormatter(formatter)

# Create the logger instance
logger = logging.getLogger('eduvora_python_worker')
logger.setLevel(logging.INFO)
logger.addHandler(info_handler)
logger.addHandler(error_handler)
logger.addHandler(console_handler)

# Prevent log propagation to the root logger to avoid duplicate console outputs
logger.propagate = False
